utils/mgmt_system/openstack.py

```diff
@@ -489,8 +490,42 @@ def get_vm_name_from_ip(self, ip):
     def get_ip_address(self, name, **kwargs):
         return self.current_ip_address(name)
 
+    def _generic_paginator(self, f):
+        """A generic paginator for OpenStack services
+
+        Takes a callable and recursively runs the "listing" until no more are returned
+        by sending the ```marker``` kwarg to offset the search results. We try to rollback
+        up to 10 times in the markers in case one was deleted. If we can't rollback after
+        10 times, we give up.
+        Possible improvement is to roll back in 5s or 10s, but then we have to check for
+        uniqueness and do dup removals.
+        """
+        lists = []
+        search_continues = True
+        marker = None
+        while search_continues:
+            if not lists:
+                temp_list = f()
+            else:
+                for i in range(min(10, len(lists))):
+                    list_offset = -(i + 1)
+                    marker = lists[list_offset].id
+                    try:
+                        temp_list = f(marker=marker)
+                        break
+                    except os_exceptions.BadRequest:
+                        continue
+                else:
+                    raise Exception("Could not get list, maybe mass deletion after 10 marker tries")
+            if not temp_list:
+                search_continues = False
+            else:
+                lists = lists + temp_list
```

Comment:
I'm probably guilty of this myself, but `list1 + list2` is a little opaque, I think `list1.extend(list2)` is easier on the eyes.

I'm also not a fan of not looking for something...
```python
if temp_list:
    lists.extend(temp_list)
else:
    # Not sure search_continues adds much value here
    break